aws-http-nio-client-netty/src/main/java/software/amazon/awssdk/http/nio/netty/internal/SharedEventLoopGroup.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.nio.netty.internal;
+
+import io.netty.channel.EventLoopGroup;
+import io.netty.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+import software.amazon.awssdk.annotation.SdkInternalApi;
+import software.amazon.awssdk.annotation.SdkTestInternalApi;
+import software.amazon.awssdk.http.nio.netty.DefaultEventLoopGroupFactory;
+
+/**
+ * Provides access and manages a shared {@link EventLoopGroup}. Uses reference counting to keep track of how many HTTP
+ * clients are using the shared event loop group and will automatically close it when that count reaches zero. Event loop
+ * group is lazily initialized for the first time and and subsequent requests after the count reaches zero.
+ */
+public class SharedEventLoopGroup {
+
+    /**
+     * Lazily initialized shared default event loop group.
+     */
+    private static EventLoopGroup sharedEventLoopGroup;
+
+    /**
+     * Reference count of clients using the shared event loop group.
+     */
+    private static int referenceCount = 0;
+
+    /**
+     * @return The default {@link EventLoopGroup} that will be shared across all service clients. This is used when the customer
+     * does not specify a custom {@link EventLoopGroup} or {@link DefaultEventLoopGroupFactory}.
+     */
+    @SdkInternalApi
+    public static synchronized EventLoopGroup get() {
+        if (sharedEventLoopGroup == null) {
+            sharedEventLoopGroup = DefaultEventLoopGroupFactory.builder().build().create();
+        }
+        referenceCount++;
+        return new ReferenceCountingEventLoopGroup(sharedEventLoopGroup);
+    }
+
+    /**
+     * Decrement the reference count and close if necessary.
+     */
+    private static synchronized void decrementReference() {
+        referenceCount--;
+        if (referenceCount == 0) {
+            sharedEventLoopGroup.shutdownGracefully();
+            sharedEventLoopGroup = null;
+        }
+    }
+
+    @SdkTestInternalApi
+    static synchronized int referenceCount() {
```

Comment:
I don't love having test-only methods, but since this is static there's probably not a 'nice' way to do the assertion.

What if you had a static setter for the `EventLoopGroupFactory` that allows you to override that and provider your own (test) implementation that you can then spy on to see if its actually been shutdown. As it stands right now you're not actually testing the shutdown logic.

I'm not sure what visibility we'd want it (probably package-private for now) but we *might* get requests to expose this so customers get the shared event loop group behaviour but with their own factory implementation.